File: apps/windows/src/conditionalPlan.ts
/**
 * Maps OpportunityRow + compact price_path → ConditionalPlan for Dashboard 2.0.
 * Pure presentation logic — no React, no I/O.
 *
 * SNR rules:
 * - ScaleIn is rare (partial entry only when score + zone conf are solid and risks are light).
 * - Mixed / weak Watch setups default to WaitZone, not ScaleIn spam.
 * - Low zone confidence is not shown as a precise $ band (avoids false precision).
 */
import type {
  CompactPricePath,
  Decision,
  OpportunityRow,
  PathMotiveCode,
  SetupLabel,
  ZoneConfidence,
} from "./api.ts";
import type { ScoringModelId } from "./scoringPresentation.ts";
import { isTechnicalAdverseForSide } from "./technicalVerdict.ts";

export type PlanStance = "ActNow" | "ScaleIn" | "WaitZone" | "Avoid";
export type PlanSide = "long" | "short";

export interface PlanEvidence {
  code: PathMotiveCode | "score" | "decision";
  textKey: string;
  vars?: Record<string, string | number>;
}

export interface ConditionalPlan {
  symbol: string;
  companyName: string | null;
  side: PlanSide;
  stance: PlanStance;
  zoneLowCents: number | null;
  zoneHighCents: number | null;
  zoneConfidence: ZoneConfidence | null;
  /** True when zone band is shown (med/high conf only). */
  zoneShown: boolean;
  pTouch20d: number | null;
  expectedSessions: number | null;
  invalidationCents: number | null;
  headlineKey: string;
  headlineVars: Record<string, string | number>;
  support: PlanEvidence[];
  caution: PlanEvidence[];
  urgency: number;
  /** Higher = clearer signal for priority list. */
  signalClarity: number;
  compositeScore: number;
  decision: Decision;
  setupLabel: SetupLabel;
  marketPriceCents: number;
  spark: number[];
  timingMethod: CompactPricePath["timing_method"] | null;
}

/** Path risks that block "enter now / scale" more than soft regime noise. */
const MATERIAL_RISKS: ReadonlySet<PathMotiveCode> = new Set([
  "extension",
  "far_from_support",
  "far_from_resistance",
  "rsi_rich",
  "rsi_washed",
  "trend_against",
  "earnings_soon",
  "weak_forecast",
  "above_value",
]);

const STRONG_SETUP: SetupLabel[] = ["StrongBuy", "Buy", "StrongAccumulate"];
const POSITIVE_SETUP: SetupLabel[] = [
  "StrongBuy",
  "Buy",
  "Accumulate",
  "StrongAccumulate",
];

export function planSideFromModel(model: ScoringModelId): PlanSide {
  return model === "short_v3" ? "short" : "long";
}

export function buildConditionalPlan(
  row: OpportunityRow,
  model: ScoringModelId,
): ConditionalPlan {
  const side = planSideFromModel(model);
  const path = row.price_path ?? null;
  const rawStance = deriveStance(row, path);
  // Same technical_score the detail panel / buckets show — never Act long into Strong Bearish.
  const stance = applyTechnicalConsistency(rawStance, row.technical_score, side);
  const zoneShown = shouldShowZone(path);
  const rawCaution = motivesToEvidence(path?.risk_codes ?? []);
  const rawSupport = motivesToEvidence(path?.support_codes ?? []);
  const techCaution = technicalCautionEvidence(row.technical_score, side);
  const mergedCaution =
    techCaution != null ? [techCaution, ...rawCaution] : rawCaution;
  const { caution, support } = pickEvidence(
    stance,
    mergedCaution,
    rawSupport,
    row.composite_score,
  );

  const { headlineKey, headlineVars } = buildHeadline(row, path, stance, side, zoneShown);
  const signalClarity = computeSignalClarity(row, path, stance, zoneShown);
  const urgency = computeUrgency(row, path, stance, zoneShown, signalClarity);

  return {
    symbol: row.symbol,
    companyName: row.company_name,
    side,
    stance,
    zoneLowCents: zoneShown ? (path?.zone_low_cents ?? null) : null,
    zoneHighCents: zoneShown ? (path?.zone_high_cents ?? null) : null,
    zoneConfidence: path?.zone_confidence ?? null,
    zoneShown,
    pTouch20d: path?.p_touch_20d ?? null,
    expectedSessions: path?.expected_sessions ?? null,
    invalidationCents: path?.invalidation_cents ?? null,
    headlineKey,
    headlineVars,
    support,
    caution,
    urgency,
    signalClarity,
    compositeScore: row.composite_score,
    decision: row.decision,
    setupLabel: row.setup_label,
    marketPriceCents: row.market_price_cents,
    spark: row.spark ?? [],
    timingMethod: path?.timing_method ?? null,
  };
}

/**
 * Align plan stance with the single technical engine score.
 * Mild adverse tech → never Act/Scale; strong adverse → Avoid.
 */
export function applyTechnicalConsistency(
  stance: PlanStance,
  technicalScore: number | null | undefined,
  side: PlanSide,
): PlanStance {
  const adverse = isTechnicalAdverseForSide(technicalScore, side);
  if (adverse == null) return stance;
  if (adverse === "strong") return "Avoid";
  // mild: demote actionable stances only
  if (stance === "ActNow" || stance === "ScaleIn") return "WaitZone";
  return stance;
}

function technicalCautionEvidence(
  technicalScore: number | null | undefined,
  side: PlanSide,
): PlanEvidence | null {
  const adverse = isTechnicalAdverseForSide(technicalScore, side);
  if (adverse == null || technicalScore == null) return null;
  return {
    code: "score",
    textKey:
      adverse === "strong"
        ? "dash.v2.ev.techStrongAdverse"
        : "dash.v2.ev.techMildAdverse",
    vars: { tech: technicalScore },
  };
}

/** Primary board: long/short entries you can act on (or scale into). Never Wait. */
export function isActionablePriority(plan: ConditionalPlan): boolean {
  if (plan.stance === "ActNow") {
    return plan.compositeScore >= 25 || plan.zoneShown;
  }
  if (plan.stance === "ScaleIn") {
    if (plan.compositeScore < 28) return false;
    if (!plan.zoneShown) return false;
    if (plan.zoneConfidence === "low") return false;
    if (plan.caution.length >= 2) return false;
    return true;
  }
  return false;
}

/**
 * Secondary radar: solid setups to re-check later — not shown as the main board.
 * Higher bar than old “everything is wait”.
 */
export function isWaitPriority(plan: ConditionalPlan): boolean {
  if (plan.stance !== "WaitZone") return false;
  if (plan.compositeScore < 35) return false;
  if (!plan.zoneShown) return false;
  if (plan.zoneConfidence !== "high" && plan.zoneConfidence !== "med") return false;
  return true;
}

/** @deprecated prefer isActionablePriority / isWaitPriority */
export function isPriorityPlan(plan: ConditionalPlan): boolean {
  return isActionablePriority(plan) || isWaitPriority(plan);
}

/** Human review horizon from expected sessions (for wait copy). */
export function reviewHorizonLabel(sessions: number | null | undefined): string {
  if (sessions == null || !Number.isFinite(sessions)) return "unas semanas";
  const s = Math.max(0, Math.round(sessions));
  if (s <= 2) return "1–2 días de sesión";
  if (s <= 5) return `${s} días de sesión`;
  if (s <= 12) return "1–2 semanas";
  if (s <= 25) return "2–4 semanas";
  if (s <= 45) return "1–2 meses";
  return "varios meses";
}

export function deriveStance(
  row: OpportunityRow,
  path: CompactPricePath | null,
): PlanStance {
  if (row.decision === "Avoid" || row.setup_label === "StrongAvoid") {
    return "Avoid";
  }
  if (row.composite_score < 0) {
    return "Avoid";
  }

  const riskCodes = path?.risk_codes ?? [];
  const material = riskCodes.filter((c) => MATERIAL_RISKS.has(c)).length;
  const softOnly =
    riskCodes.length > 0 && riskCodes.every((c) => !MATERIAL_RISKS.has(c));
  const inZone = path?.support_codes?.includes("in_zone") ?? false;
  const nearZone = path?.support_codes?.includes("near_zone") ?? false;
  const inOrNear = inZone || nearZone;
  const far =
    riskCodes.includes("far_from_support") ||
    riskCodes.includes("far_from_resistance") ||
    riskCodes.includes("extension") ||
    riskCodes.includes("rsi_rich");
  const zoneConf = path?.zone_confidence ?? null;
  const solidZone = zoneConf === "high" || zoneConf === "med";

  if (row.decision === "Act") {
    // Clean entry: in/near zone, no material path risk
    if (inOrNear && material === 0) return "ActNow";
    // One soft risk (e.g. regime) only — still act if strong setup
    if (inOrNear && material === 0 && softOnly && STRONG_SETUP.includes(row.setup_label)) {
      return "ActNow";
    }
    // Stretched / multiple material risks → wait
    if (far || material >= 2) return "WaitZone";
    // Partial: near zone, single material risk, solid zone band → rare ScaleIn
    if (nearZone && material === 1 && solidZone && row.composite_score >= 30) {
      return "ScaleIn";
    }
    // In zone but one material caution → wait for cleaner tape, not "scale"
    if (inZone && material >= 1) return "WaitZone";
    // Strong label, not extended
    if (STRONG_SETUP.includes(row.setup_label) && !far && material <= 1) {
      return material === 0 ? "ActNow" : "WaitZone";
    }
    return "WaitZone";
  }

  // Watch — default is Wait, not Scale
  if (row.composite_score < 15 && !POSITIVE_SETUP.includes(row.setup_label)) {
    return "WaitZone";
  }
  if (far || material >= 1) return "WaitZone";
  if (inOrNear && solidZone && row.composite_score >= 28 && material === 0) {
    // Only scale when score is solid and path is clean
    return "ScaleIn";
  }
  if (inOrNear && material === 0 && STRONG_SETUP.includes(row.setup_label) && solidZone) {
    return "ScaleIn";
  }
  return "WaitZone";
}

function shouldShowZone(path: CompactPricePath | null): boolean {
  if (path?.zone_low_cents == null || path?.zone_high_cents == null) return false;
  return path.zone_confidence === "med" || path.zone_confidence === "high";
}

function motivesToEvidence(codes: PathMotiveCode[]): PlanEvidence[] {
  return codes.map((code) => ({
    code,
    textKey: `dash.v2.motive.${code}`,
  }));
}

/** Prefer evidence that supports the stance; cap mixed noise at 3. */
function pickEvidence(
  stance: PlanStance,
  caution: PlanEvidence[],
  support: PlanEvidence[],
  composite: number,
): { caution: PlanEvidence[]; support: PlanEvidence[] } {
  if (stance === "ActNow") {
    const s = support.slice(0, 2);
    const c = caution.slice(0, 1);
    if (s.length === 0) {
      s.push({
        code: "score",
        textKey: "dash.v2.ev.composite",
        vars: { score: composite },
      });
    }
    return { caution: c, support: s };
  }
  if (stance === "WaitZone") {
    const c = caution.slice(0, 2);
    const s = support.slice(0, 1);
    if (c.length === 0) {
      c.push({ code: "decision", textKey: "dash.v2.ev.timingWeak" });
    }
    return { caution: c, support: s };
  }
  if (stance === "ScaleIn") {
    return {
      caution: caution.slice(0, 1),
      support: support.slice(0, 2),
    };
  }
  return { caution: caution.slice(0, 2), support: support.slice(0, 1) };
}

function buildHeadline(
  row: OpportunityRow,
  path: CompactPricePath | null,
  stance: PlanStance,
  side: PlanSide,
  zoneShown: boolean,
): { headlineKey: string; headlineVars: Record<string, string | number> } {
  const zone =
    zoneShown && path?.zone_low_cents != null && path?.zone_high_cents != null
      ? formatZone(path.zone_low_cents, path.zone_high_cents)
      : "";
  const p20 = path?.p_touch_20d;
  const inv =
    path?.invalidation_cents != null
      ? formatDollars(path.invalidation_cents)
      : "";

  const vars: Record<string, string | number> = {
    symbol: row.symbol,
    zone: zone || "—",
    p20: p20 ?? "—",
    inv: inv || "—",
    sessions: path?.expected_sessions ?? "—",
    review: reviewHorizonLabel(path?.expected_sessions),
  };

  const prefix = side === "short" ? "short." : "";
  switch (stance) {
    case "ActNow":
      return {
        headlineKey: zone
          ? `dash.v2.head.${prefix}actZone`
          : `dash.v2.head.${prefix}act`,
        headlineVars: vars,
      };
    case "ScaleIn":
      return {
        headlineKey: zone
          ? `dash.v2.head.${prefix}scaleZone`
          : `dash.v2.head.${prefix}scale`,
        headlineVars: vars,
      };
    case "WaitZone":
      return {
        headlineKey: zone
          ? `dash.v2.head.${prefix}waitReview`
          : `dash.v2.head.${prefix}wait`,
        headlineVars: vars,
      };
    case "Avoid":
      return {
        headlineKey: `dash.v2.head.${prefix}avoid`,
        headlineVars: vars,
      };
  }
}

function computeSignalClarity(
  row: OpportunityRow,
  path: CompactPricePath | null,
  stance: PlanStance,
  zoneShown: boolean,
): number {
  let c = 0;
  if (stance === "ActNow") c += 40;
  else if (stance === "WaitZone") c += 25;
  else if (stance === "ScaleIn") c += 10;
  else c -= 20;

  if (zoneShown && path?.zone_confidence === "high") c += 20;
  else if (zoneShown && path?.zone_confidence === "med") c += 10;
  else if (path?.zone_confidence === "low") c -= 25;

  const material = (path?.risk_codes ?? []).filter((x) => MATERIAL_RISKS.has(x)).length;
  if (stance === "ActNow" && material === 0) c += 15;
  if (stance === "WaitZone" && material >= 1) c += 10;
  // Mixed: support says in_zone but many material risks
  if ((path?.support_codes?.includes("in_zone") ?? false) && material >= 2) c -= 20;

  if (row.composite_score >= 40) c += 15;
  else if (row.composite_score >= 30) c += 8;
  else if (row.composite_score < 15) c -= 15;

  if (row.confidence === "High") c += 5;
  if (row.confidence === "Low") c -= 8;

  return c;
}

function computeUrgency(
  row: OpportunityRow,
  path: CompactPricePath | null,
  stance: PlanStance,
  zoneShown: boolean,
  signalClarity: number,
): number {
  let u = row.composite_score + signalClarity;

  if (stance === "ActNow") u += 50;
  else if (stance === "WaitZone") u += 20;
  else if (stance === "ScaleIn") u += 5;
  else u -= 40;

  if (path?.support_codes?.includes("in_zone") && stance === "ActNow") u += 30;
  if (path?.support_codes?.includes("near_zone") && stance === "WaitZone") u += 15;

  if (zoneShown && path?.zone_confidence === "high") u += 12;
  if (!zoneShown && path?.zone_confidence === "low") u -= 30;

  return u;
}

export function formatDollars(cents: number): string {
  return `$${(cents / 100).toFixed(2)}`;
}

export function formatZone(low: number, high: number): string {
  return `${formatDollars(low)}–${formatDollars(high)}`;
}

File: apps/android/core/src/main/kotlin/com/discountscreener/core/plan/ConditionalPlan.kt
package com.discountscreener.core.plan

import com.discountscreener.core.engine.OpportunityEngine
import com.discountscreener.core.model.OpportunityScoringModel

enum class PlanStance {
    ActNow,
    ScaleIn,
    WaitZone,
    Avoid,
}

enum class PlanDecision {
    Act,
    Watch,
    Avoid,
}

enum class SetupLabel {
    StrongBuy,
    Buy,
    StrongAccumulate,
    Accumulate,
    Watch,
    HoldWait,
    Hold,
    Neutral,
    Caution,
    Distribute,
    Avoid,
    StrongAvoid,
}

data class PlanEvidence(
    val code: String,
    val text: String,
)

data class ConditionalPlan(
    val symbol: String,
    val companyName: String?,
    val stance: PlanStance,
    val zoneLowCents: Long?,
    val zoneHighCents: Long?,
    val zoneConfidence: ZoneConfidence?,
    val zoneShown: Boolean,
    val pTouch20d: Int?,
    val expectedSessions: Int?,
    val invalidationCents: Long?,
    val headline: String,
    val support: List<PlanEvidence>,
    val caution: List<PlanEvidence>,
    val urgency: Int,
    val signalClarity: Int,
    val compositeScore: Int,
    val technicalScore: Int?,
    val technicalSignals: List<String>,
    val decision: PlanDecision,
    val setupLabel: SetupLabel,
    val marketPriceCents: Long,
    val spark: List<Long>,
    val timingMethod: TimingMethod?,
)

data class PlanRowInput(
    val symbol: String,
    val companyName: String? = null,
    val decision: PlanDecision,
    val compositeScore: Int,
    val technicalScore: Int?,
    val technicalSignals: List<String> = emptyList(),
    val forecastScore: Int? = null,
    val confidenceHigh: Boolean = false,
    val marketPriceCents: Long,
    val streetFairValueCents: Long = 0,
    val dcfValueCents: Long? = null,
    val analystLowCents: Long? = null,
    val gapBps: Int? = null,
    val path: CompactPricePath? = null,
    val spark: List<Long> = emptyList(),
    val scoringModel: OpportunityScoringModel = OpportunityScoringModel.AggressiveV4,
    val regimeStance: String? = null,
)

private val MATERIAL_RISKS: Set<PathMotiveCode> = setOf(
    PathMotiveCode.Extension,
    PathMotiveCode.FarFromSupport,
    PathMotiveCode.RsiRich,
    PathMotiveCode.RsiWashed,
    PathMotiveCode.TrendAgainst,
    PathMotiveCode.EarningsSoon,
    PathMotiveCode.WeakForecast,
    PathMotiveCode.AboveValue,
)

private val STRONG_SETUP: Set<SetupLabel> = setOf(SetupLabel.StrongBuy, SetupLabel.Buy, SetupLabel.StrongAccumulate)
private val POSITIVE_SETUP: Set<SetupLabel> = STRONG_SETUP + SetupLabel.Accumulate

private val HOSTILE_REGIME: Set<String> = setOf(
    "HoldTrim",
    "Reduce",
    "Defend",
    "Denial",
    "UnstableBlowoff",
    "Euphoria",
    "Distribute",
)

fun setupFromComposite(composite: Int): SetupLabel = when {
    composite >= 50 -> SetupLabel.StrongBuy
    composite >= 30 -> SetupLabel.Buy
    composite >= 15 -> SetupLabel.Accumulate
    composite >= 0 -> SetupLabel.Watch
    composite >= -20 -> SetupLabel.Hold
    composite >= -40 -> SetupLabel.Avoid
    else -> SetupLabel.StrongAvoid
}

fun buildConditionalPlan(row: PlanRowInput): ConditionalPlan {
    var path = row.path
    var setup = setupFromComposite(row.compositeScore)
    var rawStance = deriveStance(row.decision, row.compositeScore, setup, path)
    var stance = applyTechnicalConsistency(rawStance, row.technicalScore)
    var zoneShown = shouldShowZone(path)
    var rawCaution = motivesToEvidence(path?.riskCodes.orEmpty())
    var rawSupport = motivesToEvidence(path?.supportCodes.orEmpty())
    var techCaution = technicalCautionEvidence(row.technicalScore)
    var mergedCaution = if (techCaution != null) listOf(techCaution) + rawCaution else rawCaution
    var picked = pickEvidence(stance, mergedCaution, rawSupport, row.compositeScore)
    var caution = picked.first
    var support = picked.second
    var gate = ActionableGateInput(
        stance = stance,
        technicalScore = row.technicalScore,
        technicalSignals = row.technicalSignals,
        compositeScore = row.compositeScore,
        zoneShown = zoneShown,
        zoneConfidence = path?.zoneConfidence,
        cautionCount = caution.size,
        scoringModel = row.scoringModel,
    )
    if ((stance == PlanStance.ActNow || stance == PlanStance.ScaleIn) && !passesActionableGates(gate)) {
        stance = PlanStance.WaitZone
        picked = pickEvidence(stance, mergedCaution, rawSupport, row.compositeScore)
        caution = picked.first
        support = picked.second
    }
    if (stance == PlanStance.ActNow && shouldDemoteForRegime(row, path, zoneShown)) {
        stance = PlanStance.WaitZone
        picked = pickEvidence(stance, mergedCaution, rawSupport, row.compositeScore)
        caution = picked.first
        support = picked.second
        if (caution.none { it.code == "regime" }) {
            caution = listOf(PlanEvidence("regime", PlanCopy.motive(PathMotiveCode.RegimeRisk))) + caution
            caution = caution.take(2)
        }
    }
    var p20 = publishableP20(path?.timingMethod ?: TimingMethod.Unavailable, path?.pTouch20d)
    var headline = PlanCopy.headline(
        stance = stance,
        zone = if (zoneShown) formatZone(path?.zoneLowCents, path?.zoneHighCents) else null,
        p20 = p20,
        inv = path?.invalidationCents?.let { formatDollars(it) },
        review = reviewHorizonLabel(path?.expectedSessions),
    )
    var clarity = computeSignalClarity(row, path, stance, zoneShown)
    var urgency = computeUrgency(row, path, stance, zoneShown, clarity)
    return ConditionalPlan(
        symbol = row.symbol,
        companyName = row.companyName,
        stance = stance,
        zoneLowCents = if (zoneShown) path?.zoneLowCents else null,
        zoneHighCents = if (zoneShown) path?.zoneHighCents else null,
        zoneConfidence = path?.zoneConfidence,
        zoneShown = zoneShown,
        pTouch20d = p20,
        expectedSessions = path?.expectedSessions,
        invalidationCents = path?.invalidationCents,
        headline = headline,
        support = support,
        caution = caution,
        urgency = urgency,
        signalClarity = clarity,
        compositeScore = row.compositeScore,
        technicalScore = row.technicalScore,
        technicalSignals = row.technicalSignals,
        decision = row.decision,
        setupLabel = setup,
        marketPriceCents = row.marketPriceCents,
        spark = row.spark,
        timingMethod = path?.timingMethod,
    )
}

fun applyTechnicalConsistency(stance: PlanStance, technicalScore: Int?): PlanStance {
    var adverse = technicalAdverse(technicalScore)
    if (adverse == null) return stance
    if (adverse == "strong") return PlanStance.Avoid
    if (stance == PlanStance.ActNow || stance == PlanStance.ScaleIn) return PlanStance.WaitZone
    return stance
}

fun technicalAdverse(technicalScore: Int?): String? {
    if (technicalScore == null) return null
    if (technicalScore <= -45) return "strong"
    if (technicalScore <= -15) return "mild"
    return null
}

fun hasStructuralConflict(signals: List<String>): Boolean =
    signals.contains("50/200-") || signals.contains("20/50-")

data class ActionableGateInput(
    val stance: PlanStance,
    val technicalScore: Int?,
    val technicalSignals: List<String>,
    val compositeScore: Int,
    val zoneShown: Boolean,
    val zoneConfidence: ZoneConfidence?,
    val cautionCount: Int,
    val scoringModel: OpportunityScoringModel,
)

fun passesActionableGates(plan: ActionableGateInput): Boolean {
    if (plan.technicalScore == null) return false
    if (!hasActionableTechnicalFloor(plan.technicalScore)) return false
    if (hasStructuralConflict(plan.technicalSignals) && !hasStrongTechnical(plan.technicalScore)) return false
    var actFloor = OpportunityEngine.actAtOrAboveScore(plan.scoringModel)
    if (plan.stance == PlanStance.ActNow) {
        return plan.compositeScore >= actFloor && plan.zoneShown
    }
    if (plan.stance == PlanStance.ScaleIn) {
        if (plan.compositeScore < 28) return false
        if (!plan.zoneShown) return false
        if (plan.zoneConfidence == ZoneConfidence.Low) return false
        if (plan.cautionCount >= 2) return false
        return true
    }
    return false
}

fun isActionablePriority(plan: ConditionalPlan, scoringModel: OpportunityScoringModel): Boolean =
    passesActionableGates(
        ActionableGateInput(
            stance = plan.stance,
            technicalScore = plan.technicalScore,
            technicalSignals = plan.technicalSignals,
            compositeScore = plan.compositeScore,
            zoneShown = plan.zoneShown,
            zoneConfidence = plan.zoneConfidence,
            cautionCount = plan.caution.size,
            scoringModel = scoringModel,
        ),
    )

fun isWaitPriority(plan: ConditionalPlan): Boolean {
    if (plan.stance != PlanStance.WaitZone) return false
    if (plan.compositeScore < 35) return false
    if (!plan.zoneShown) return false
    if (plan.zoneConfidence != ZoneConfidence.High && plan.zoneConfidence != ZoneConfidence.Med) return false
    return true
}

fun deriveStance(
    decision: PlanDecision,
    compositeScore: Int,
    setup: SetupLabel,
    path: CompactPricePath?,
): PlanStance {
    if (decision == PlanDecision.Avoid || setup == SetupLabel.StrongAvoid) return PlanStance.Avoid
    if (compositeScore < 0) return PlanStance.Avoid
    var riskCodes = path?.riskCodes.orEmpty()
    var material = riskCodes.count { MATERIAL_RISKS.contains(it) }
    var inZone = path?.supportCodes.orEmpty().contains(PathMotiveCode.InZone)
    var nearZone = path?.supportCodes.orEmpty().contains(PathMotiveCode.NearZone)
    var inOrNear = inZone || nearZone
    var far = riskCodes.contains(PathMotiveCode.FarFromSupport) ||
        riskCodes.contains(PathMotiveCode.Extension) ||
        riskCodes.contains(PathMotiveCode.RsiRich)
    var zoneConf = path?.zoneConfidence
    var solidZone = zoneConf == ZoneConfidence.High || zoneConf == ZoneConfidence.Med
    if (decision == PlanDecision.Act) {
        if (inOrNear && material == 0) return PlanStance.ActNow
        if (far || material >= 2) return PlanStance.WaitZone
        if (nearZone && material == 1 && solidZone && compositeScore >= 30) return PlanStance.ScaleIn
        if (inZone && material >= 1) return PlanStance.WaitZone
        if (STRONG_SETUP.contains(setup) && !far && material <= 1) {
            return if (material == 0) PlanStance.ActNow else PlanStance.WaitZone
        }
        return PlanStance.WaitZone
    }
    if (compositeScore < 15 && !POSITIVE_SETUP.contains(setup)) return PlanStance.WaitZone
    if (far || material >= 1) return PlanStance.WaitZone
    if (inOrNear && solidZone && compositeScore >= 28 && material == 0) return PlanStance.ScaleIn
    if (inOrNear && material == 0 && STRONG_SETUP.contains(setup) && solidZone) return PlanStance.ScaleIn
    return PlanStance.WaitZone
}

fun reviewHorizonLabel(sessions: Int?): String {
    if (sessions == null) return "a few weeks"
    var s = sessions.coerceAtLeast(0)
    return when {
        s <= 2 -> "1–2 sessions"
        s <= 5 -> "$s sessions"
        s <= 12 -> "1–2 weeks"
        s <= 25 -> "2–4 weeks"
        s <= 45 -> "1–2 months"
        else -> "several months"
    }
}

fun formatDollars(cents: Long): String = "$%.2f".format(cents / 100.0)

fun formatZone(low: Long?, high: Long?): String? {
    if (low == null || high == null) return null
    return "${formatDollars(low)}–${formatDollars(high)}"
}

private fun shouldShowZone(path: CompactPricePath?): Boolean {
    if (path?.zoneLowCents == null || path.zoneHighCents == null) return false
    return path.zoneConfidence == ZoneConfidence.Med || path.zoneConfidence == ZoneConfidence.High
}

private fun hasActionableTechnicalFloor(tech: Int): Boolean = tech >= 10

private fun hasStrongTechnical(tech: Int): Boolean = tech >= 30

private fun shouldDemoteForRegime(row: PlanRowInput, path: CompactPricePath?, zoneShown: Boolean): Boolean {
    var stance = row.regimeStance ?: return false
    if (!HOSTILE_REGIME.contains(stance)) return false
    var inZone = path?.supportCodes.orEmpty().contains(PathMotiveCode.InZone)
    var highConf = path?.zoneConfidence == ZoneConfidence.High
    if (inZone && highConf && zoneShown && row.compositeScore >= 40) return false
    return true
}

private fun motivesToEvidence(codes: List<PathMotiveCode>): List<PlanEvidence> =
    codes.map { PlanEvidence(it.name, PlanCopy.motive(it)) }

private fun technicalCautionEvidence(technicalScore: Int?): PlanEvidence? {
    var adverse = technicalAdverse(technicalScore) ?: return null
    if (technicalScore == null) return null
    var text = if (adverse == "strong") {
        PlanCopy.techStrongAdverse(technicalScore)
    } else {
        PlanCopy.techMildAdverse(technicalScore)
    }
    return PlanEvidence("score", text)
}

private fun pickEvidence(
    stance: PlanStance,
    caution: List<PlanEvidence>,
    support: List<PlanEvidence>,
    composite: Int,
): Pair<List<PlanEvidence>, List<PlanEvidence>> {
    if (stance == PlanStance.ActNow) {
        var s = support.take(2).toMutableList()
        var c = caution.take(1)
        if (s.isEmpty()) s.add(PlanEvidence("score", PlanCopy.composite(composite)))
        return c to s
    }
    if (stance == PlanStance.WaitZone) {
        var c = caution.take(2).toMutableList()
        var s = support.take(1)
        if (c.isEmpty()) c.add(PlanEvidence("decision", PlanCopy.timingWeak()))
        return c to s
    }
    if (stance == PlanStance.ScaleIn) {
        return caution.take(1) to support.take(2)
    }
    return caution.take(2) to support.take(1)
}

private fun computeSignalClarity(
    row: PlanRowInput,
    path: CompactPricePath?,
    stance: PlanStance,
    zoneShown: Boolean,
): Int {
    var c = 0
    c += when (stance) {
        PlanStance.ActNow -> 40
        PlanStance.WaitZone -> 25
        PlanStance.ScaleIn -> 10
        PlanStance.Avoid -> -20
    }
    if (zoneShown && path?.zoneConfidence == ZoneConfidence.High) c += 20
    else if (zoneShown && path?.zoneConfidence == ZoneConfidence.Med) c += 10
    else if (path?.zoneConfidence == ZoneConfidence.Low) c -= 25
    var material = path?.riskCodes.orEmpty().count { MATERIAL_RISKS.contains(it) }
    if (stance == PlanStance.ActNow && material == 0) c += 15
    if (stance == PlanStance.WaitZone && material >= 1) c += 10
    if (path?.supportCodes.orEmpty().contains(PathMotiveCode.InZone) && material >= 2) c -= 20
    c += when {
        row.compositeScore >= 40 -> 15
        row.compositeScore >= 30 -> 8
        row.compositeScore < 15 -> -15
        else -> 0
    }
    if (row.confidenceHigh) c += 5
    return c
}

private fun computeUrgency(
    row: PlanRowInput,
    path: CompactPricePath?,
    stance: PlanStance,
    zoneShown: Boolean,
    signalClarity: Int,
): Int {
    var u = row.compositeScore + signalClarity
    u += when (stance) {
        PlanStance.ActNow -> 50
        PlanStance.WaitZone -> 20
        PlanStance.ScaleIn -> 5
        PlanStance.Avoid -> -40
    }
    if (path?.supportCodes.orEmpty().contains(PathMotiveCode.InZone) && stance == PlanStance.ActNow) u += 30
    if (path?.supportCodes.orEmpty().contains(PathMotiveCode.NearZone) && stance == PlanStance.WaitZone) u += 15
    if (zoneShown && path?.zoneConfidence == ZoneConfidence.High) u += 12
    if (!zoneShown && path?.zoneConfidence == ZoneConfidence.Low) u -= 30
    return u
}
